src/ui.rs

```diff
@@ -84,19 +94,67 @@ fn use_color(choice: ColorChoice) -> bool {
     }
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum PaginationChoice {
+    Always,
+    Never,
+    Auto,
+}
+
+impl Default for PaginationChoice {
+    fn default() -> Self {
+        PaginationChoice::Auto
+    }
+}
+
+impl FromStr for PaginationChoice {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "always" => Ok(PaginationChoice::Always),
+            "never" => Ok(PaginationChoice::Never),
+            "auto" => Ok(PaginationChoice::Auto),
+            _ => Err("must be one of always, never, or auto"),
+        }
+    }
+}
+
+fn pagination_setting(settings: &UserSettings) -> PaginationChoice {
+    settings
+        .config()
+        .get_string("ui.paginate")
+        .ok()
+        .and_then(|s| s.parse().ok())
+        .unwrap_or_default()
+}
+
+fn pager_setting(settings: &UserSettings) -> String {
+    settings
+        .config()
+        .get_string("ui.pager")
+        .unwrap_or_else(|_| "less".to_string())
+}
+
 impl Ui {
     pub fn for_terminal(settings: UserSettings) -> Ui {
         let cwd = std::env::current_dir().unwrap();
         let color = use_color(color_setting(&settings));
+        let paginate = pagination_setting(&settings);
+        let progress_indicator = progress_indicator_setting(&settings);
         let formatter_factory = FormatterFactory::prepare(&settings, color);
+        let output = match paginate {
+            PaginationChoice::Always => UiOutput::new_paged_else_terminal(&settings),
```

Comment:
I don't think we should direct output to a pager unconditionally on startup. Long-running operations might want to display progress feedback directly to the terminal before pager-appropriate data is available, for example. I instead interpret "paginate always" the same way as "use color always", i.e. meaning "regardless of whether stdout looks like a tty."

Reply:
Which is to say, we should still defer spawning the pager until requested.

Reply:
Heh, I thought the same thing (I think I may have told @chooglen in person). But does it make sense to page when stdout doesn't look like a tty? Are there cases where that's useful?

Reply:
> I instead interpret "paginate always" the same way as "use color always", i.e. meaning "regardless of whether stdout looks like a tty."

Ah okay. I think the main issue (and the one I'm trying to address) is "what is the set of commands that trigger pagination?", which makes me think that I should just rename the config option to match that.

Reply:
Oh, and I also checked that `hg` does the same thing. So does `git`:
```
$ PAGER='echo foo' git -p rm foo
foo
```
It doesn't even mention the error (there's no `foo` file in my repo)

Reply:
> Ah okay. I think the main issue (and the one I'm trying to address) is "what is the set of commands that trigger pagination?", which makes me think that I should just rename the config option to match that.

Actually, this is probably better done via a CLI flag. I'll send that in a follow-up PR.

Reply:
What would that flag be? Whatever it is, I agree to leave it for later so we the base feature in.

Reply:
In the follow up PR, I'm thinking of adding `--[no-]pager`, which is a bit like [hg](https://www.mercurial-scm.org/doc/hg.1.html). `hg` also accepts `--pager=(always|never|auto)` (which is what the config option is based on), but I'll drop it since it's confusingly similar-sounding to `ui.color`, but uses completely different rules.

Reply:
> what is the set of commands that trigger pagination?

I'd be surprised for pagination to ever be applied to code that doesn't have specific hardcoded support for it. Paginating terse status feedback would be strange.

> But does it make sense to page when stdout doesn't look like a tty?

I speculate that there's cases where stdout isn't technically a tty but something pager-like might still be useful. Perhaps a graphical pager when `jj` is invoked as a subprocess? Or perhaps tty cases where stdout is being mangled in some way that the user's configured pager is smart enough to bypass. I recall we saw `less` has a bunch of tricks to find the "real" tty.

Reply:
Okay, in the latest version:

- The pager is not always created on startup, only certain commands can spawn it.
- Added a CLI option that lets users disable the pager.
- `ui.paginate` is gone. We can figure out what to do about `always|never|auto` in a later PR.